File: typescript/src/validators.ts
import { builtinAggregates } from "./aggregate.js";
import { ValidatorTypes, type ValidatorDefinition } from "./definition.js";
import { MessageResolver } from "./messageResolver.js";

/**
 * 検証のときに「その項目の値以外」が要るときの持ち物。
 *
 * 項目間の検証（`compare`）は**他の項目の値**を見る。値だけ渡す形では書けないので、
 * レコードごと渡す。ラベルも一緒に渡すのは、メッセージを**画面の言葉**で出すため
 * （「startDate 以上にしてください」ではなく「開始日 以上にしてください」）。
 *
 * 引数を増やすたびに拡張の署名が変わると、足すたびにプラグインが壊れる。だから
 * **1つのオブジェクト**にまとめてある（要るものが増えてもキーが増えるだけ）。
 */
export interface ValidatorContext {
  /** 検証しているレコード全体。 */
  record?: Record<string, unknown>;
  /** 項目名 → ラベル。 */
  labels?: Record<string, string>;
  /** `{ mode: create }` のような状態。 */
  mode?: string;
}

/**
 * Validates a value against a rule; returns an error message or null.
 *
 * 第3引数は任意（[ValidatorContext]）。今までの2引数の関数もそのまま渡せる。
 */
export type ValidatorFn = (
  value: unknown,
  def: ValidatorDefinition,
  context?: ValidatorContext,
) => string | null;

const isEmpty = (v: unknown): boolean =>
  v == null ||
  (typeof v === "string" && v.trim() === "") ||
  (Array.isArray(v) && v.length === 0);

const toNum = (v: unknown): number | null => {
  if (typeof v === "number") return v;
  if (typeof v === "string" && v.trim() !== "" && !Number.isNaN(Number(v))) {
    return Number(v);
  }
  return null;
};

/**
 * Built-in validators. Messages are resolved via a `MessageResolver` (default
 * locale Japanese) so they can be localized; per-rule override still wins.
 */
export function builtinValidators(
  messages: MessageResolver = new MessageResolver(),
): Record<string, ValidatorFn> {
  const m = messages;
  return {
    [ValidatorTypes.required]: (v) => (isEmpty(v) ? m.resolve("required") : null),
    [ValidatorTypes.maxLength]: (v, d) => {
      const max = toNum(d.params.value);
      if (max == null || v == null) return null;
      return String(v).length > max
        ? m.resolve("maxLength", { value: max })
        : null;
    },
    [ValidatorTypes.minLength]: (v, d) => {
      const min = toNum(d.params.value);
      if (min == null || isEmpty(v)) return null;
      return String(v).length < min
        ? m.resolve("minLength", { value: min })
        : null;
    },
    [ValidatorTypes.min]: (v, d) => {
      const min = toNum(d.params.value);
      const n = toNum(v);
      if (min == null || n == null) return null;
      return n < min ? m.resolve("min", { value: min }) : null;
    },
    [ValidatorTypes.max]: (v, d) => {
      const max = toNum(d.params.value);
      const n = toNum(v);
      if (max == null || n == null) return null;
      return n > max ? m.resolve("max", { value: max }) : null;
    },
    [ValidatorTypes.pattern]: (v, d) => {
      const src = d.params.pattern;
      if (typeof src !== "string" || isEmpty(v)) return null;
      return new RegExp(src).test(String(v)) ? null : m.resolve("pattern");
    },
    [ValidatorTypes.email]: (v) => {
      if (isEmpty(v)) return null;
      return /^[\w.+-]+@[\w-]+\.[\w.-]+$/.test(String(v))
        ? null
        : m.resolve("email");
    },
    [ValidatorTypes.postalCode]: (v) => {
      if (isEmpty(v)) return null;
      return /^\d{3}-?\d{4}$/.test(String(v)) ? null : m.resolve("postalCode");
    },
    [ValidatorTypes.compare]: (v, d, context) => compare(v, d, context ?? {}, m),
  };
}

/** `compare` で使える突合（大小を比べられるものだけ）。 */
export const COMPARE_OPERATORS = [
  "equals",
  "notEquals",
  "gt",
  "gte",
  "lt",
  "lte",
] as const;

/**
 * 項目間の検証（「終了日 は 開始日 以上」「合計 は 明細の和と同じ」）。
 *
 * 比べ方は**数として読めれば数、読めなければ文字**。ISO の日付（`2026-01-05`）は
 * 桁が揃っているので文字の大小＝日付の前後になるので、日付型を持ち込まずに済む
 * （3つのエディションで同じ答えを出すのが最優先。日付の解釈は言語ごとに違う）。
 *
 * 判定できないときは**通す**（黙って落とさない）:
 *   ・自分が空 … `required` の担当
 *   ・相手が空・相手の項目が無い … 相手側の検証の担当
 *   ・`field` が無い / 突合が使えない … 書き方の間違い。`hatake validate` が警告で言う
 */
function compare(
  value: unknown,
  def: ValidatorDefinition,
  context: ValidatorContext,
  messages: MessageResolver,
): string | null {
  const target = typeof def.params.field === "string" ? def.params.field : undefined;
  const operator =
    typeof def.params.operator === "string" ? def.params.operator : "gte";
  if (target === undefined || !COMPARE_OPERATORS.includes(operator as never)) {
    return null;
  }
  if (isEmpty(value)) return null;

  const record = context.record ?? {};
  const other = compareTo(record[target], def);
  if (other === undefined || isEmpty(other)) return null;

  const label = context.labels?.[target] ?? target;
  const aggregate =
    typeof def.params.aggregate === "string" ? def.params.aggregate : undefined;
  return holds(value, operator, other)
    ? null
    : messages.resolve(`compare.${operator}`, {
        target: aggregate === undefined ? label : `${label} の ${aggregate}`,
      });
}

/**
 * 比べる相手の値。`aggregate` があれば**明細を畳んだ数**（「合計＝明細の和」）。
 *
 * 畳み込みはダッシュボードと同じ実装を使う（同じ集約を2つ持たない）。
 */
function compareTo(raw: unknown, def: ValidatorDefinition): unknown {
  const aggregate =
    typeof def.params.aggregate === "string" ? def.params.aggregate : undefined;
  if (aggregate === undefined) return raw;
  const fold = builtinAggregates[aggregate];
  if (fold === undefined) return undefined;
  const rows = Array.isArray(raw)
    ? raw.filter(
        (row): row is Record<string, unknown> =>
          typeof row === "object" && row !== null && !Array.isArray(row),
      )
    : [];
  const of = typeof def.params.of === "string" ? def.params.of : undefined;
  return fold(rows, of);
}

/** 突合そのもの。数として読めれば数、読めなければ文字（ISO の日付はこれで前後が合う）。 */
function holds(value: unknown, operator: string, other: unknown): boolean {
  const left = toNum(value);
  const right = toNum(other);
  const [a, b]: [number | string, number | string] =
    left !== null && right !== null
      ? [left, right]
      : [String(value), String(other)];
  switch (operator) {
    case "equals":
      return a === b;
    case "notEquals":
      return a !== b;
    case "gt":
      return a > b;
    case "gte":
      return a >= b;
    case "lt":
      return a < b;
    default:
      return a <= b;
  }
}

/** Resolves validator types to implementations. Extensible via `register`. */
export class ValidatorRegistry {
  private readonly validators: Record<string, ValidatorFn>;

  constructor(custom?: Record<string, ValidatorFn>, messages?: MessageResolver) {
    this.validators = { ...builtinValidators(messages), ...custom };
  }

  run(
    value: unknown,
    def: ValidatorDefinition,
    context?: ValidatorContext,
  ): string | null {
    return this.validators[def.type]?.(value, def, context) ?? null;
  }

  register(type: string, fn: ValidatorFn): void {
    this.validators[type] = fn;
  }

  has(type: string): boolean {
    return type in this.validators;
  }
}

File: flutter/packages/hatake_core/lib/src/validation/validators.dart
import '../definition/validator_definition.dart';
import '../definition/validator_types.dart';
import '../i18n/message_resolver.dart';
import '../logic/aggregate.dart';

/// Validates a single [value] against a [ValidatorDefinition], returning an
/// error message on failure or null when valid.
typedef ValidatorFn = String? Function(
  Object? value,
  ValidatorDefinition definition,
);

/// What a validator needs besides the value it is checking.
///
/// Cross-field validation (`compare`) reads **another field's value**, which a
/// value-only signature cannot express, so the whole record comes along. The
/// labels come too, so the message can be phrased in the words on the screen
/// (「開始日以上にしてください」, not 「startDate 以上にしてください」).
///
/// One object rather than more parameters: adding a parameter would break every
/// plugin's validator each time something new is needed.
class ValidationContext {
  const ValidationContext({this.record, this.labels, this.mode});

  /// The record being validated.
  final Map<String, Object?>? record;

  /// Field name → label.
  final Map<String, String>? labels;

  /// State for `{ mode: create }` style conditions.
  final String? mode;
}

/// A validator that also sees the rest of the record ([ValidationContext]).
///
/// Kept separate from [ValidatorFn] so existing two-argument validators — the
/// ones apps and plugins already registered — keep compiling.
typedef ContextValidatorFn = String? Function(
  Object? value,
  ValidatorDefinition definition,
  ValidationContext context,
);

num? _toNum(Object? value) {
  if (value is num) return value;
  if (value is String) return num.tryParse(value);
  return null;
}

bool _isEmpty(Object? value) =>
    value == null ||
    (value is String && value.trim().isEmpty) ||
    (value is Iterable && value.isEmpty);

/// Built-in validators keyed by [ValidatorTypes]. Messages are resolved via a
/// [MessageResolver] (default locale Japanese) so they can be localized or
/// overridden; per-rule override via `ValidatorDefinition.message` still wins.
///
/// Every built-in takes a [ValidationContext] — most ignore it; `compare` needs
/// it (it reads another field of the same record).
Map<String, ContextValidatorFn> builtinValidators([MessageResolver? messages]) {
  final m = messages ?? MessageResolver();
  return {
    ValidatorTypes.required: (value, def, context) =>
        _isEmpty(value) ? m.resolve('required') : null,
    ValidatorTypes.maxLength: (value, def, context) {
      final max = (def.params['value'] as num?)?.toInt();
      if (max == null || value == null) return null;
      return value.toString().length > max
          ? m.resolve('maxLength', {'value': max})
          : null;
    },
    ValidatorTypes.minLength: (value, def, context) {
      final min = (def.params['value'] as num?)?.toInt();
      if (min == null || _isEmpty(value)) return null;
      return value.toString().length < min
          ? m.resolve('minLength', {'value': min})
          : null;
    },
    ValidatorTypes.min: (value, def, context) {
      final min = def.params['value'] as num?;
      final n = _toNum(value);
      if (min == null || n == null) return null;
      return n < min ? m.resolve('min', {'value': min}) : null;
    },
    ValidatorTypes.max: (value, def, context) {
      final max = def.params['value'] as num?;
      final n = _toNum(value);
      if (max == null || n == null) return null;
      return n > max ? m.resolve('max', {'value': max}) : null;
    },
    ValidatorTypes.pattern: (value, def, context) {
      final source = def.params['pattern'] as String?;
      if (source == null || _isEmpty(value)) return null;
      return RegExp(source).hasMatch(value.toString())
          ? null
          : m.resolve('pattern');
    },
    ValidatorTypes.email: (value, def, context) {
      if (_isEmpty(value)) return null;
      final re = RegExp(r'^[\w.+-]+@[\w-]+\.[\w.-]+$');
      return re.hasMatch(value.toString()) ? null : m.resolve('email');
    },
    ValidatorTypes.postalCode: (value, def, context) {
      if (_isEmpty(value)) return null;
      return RegExp(r'^\d{3}-?\d{4}$').hasMatch(value.toString())
          ? null
          : m.resolve('postalCode');
    },
    ValidatorTypes.compare: (value, def, context) =>
        _compare(value, def, context, m),
  };
}

/// Comparisons `compare` can make (only the ones with an order).
const List<String> compareOperators = [
  'equals',
  'notEquals',
  'gt',
  'gte',
  'lt',
  'lte',
];

/// Cross-field validation (「終了日 は 開始日 以上」「合計 は 明細の和と同じ」).
///
/// Compared **as numbers when both read as numbers, as text otherwise**. An ISO
/// date (`2026-01-05`) is zero-padded, so text order is date order — which keeps
/// date parsing (and its per-language differences) out of the comparison. All
/// three editions must answer the same thing; that matters more than types.
///
/// Passes when it cannot judge (never fails silently in the other direction):
/// this field empty (`required`'s job), the other field empty or absent (that
/// field's own rules), or a missing `field` / unusable operator — a writing
/// mistake, which `hatake validate` reports as a warning.
String? _compare(
  Object? value,
  ValidatorDefinition def,
  ValidationContext context,
  MessageResolver messages,
) {
  final target = def.params['field'];
  final operator = def.params['operator'] is String
      ? def.params['operator'] as String
      : 'gte';
  if (target is! String || !compareOperators.contains(operator)) return null;
  if (_isEmpty(value)) return null;

  final other = _compareTo(context.record?[target], def);
  if (other == null || _isEmpty(other)) return null;

  final label = context.labels?[target] ?? target;
  final aggregate = def.params['aggregate'];
  return _holds(value, operator, other)
      ? null
      : messages.resolve('compare.$operator', {
          'target': aggregate is String ? '$label の $aggregate' : label,
        });
}

/// The value to compare against. With `aggregate`, the child rows folded into a
/// number (「合計＝明細の和」) — using the same fold the dashboard uses.
///
/// `where` narrows the rows before folding, exactly as `computed` does — the same rows
/// filtered by the same rule (a subtotal that skips cancelled rows would never match a
/// check that does not).
Object? _compareTo(Object? raw, ValidatorDefinition def) {
  final aggregate = def.params['aggregate'];
  if (aggregate is! String) return raw;
  final rows = <Map<String, Object?>>[
    if (raw is Iterable)
      for (final row in raw)
        if (row is Map) row.cast<String, Object?>(),
  ];
  final of = def.params['of'];
  return AggregateRegistry().aggregate(
    aggregate,
    rowsMatching(rows, def.params['where']),
    field: of is String ? of : null,
  );
}

/// The comparison itself.
bool _holds(Object? value, String operator, Object? other) {
  final left = _toNum(value);
  final right = _toNum(other);
  final Comparable<Object> a =
      left != null && right != null ? left : value.toString();
  final Comparable<Object> b =
      left != null && right != null ? right : other.toString();
  final order = a.compareTo(b);
  switch (operator) {
    case 'equals':
      return order == 0;
    case 'notEquals':
      return order != 0;
    case 'gt':
      return order > 0;
    case 'gte':
      return order >= 0;
    case 'lt':
      return order < 0;
    default:
      return order <= 0;
  }
}

/// Wraps a value-only validator so it can live in the same table as the ones
/// that read the record.
ContextValidatorFn _withoutContext(ValidatorFn fn) =>
    (value, definition, context) => fn(value, definition);

/// Resolves validator types to their implementations. Extensible: register
/// custom validators (plugins) without modifying the framework.
class ValidatorRegistry {
  final Map<String, ContextValidatorFn> _validators;

  /// [custom] adds/overrides validators; [messages] localizes built-in
  /// messages (default locale Japanese).
  ///
  /// [custom] takes the two-argument [ValidatorFn] (what apps and plugins
  /// already write); they are wrapped so everything runs through one signature.
  ValidatorRegistry([
    Map<String, ValidatorFn>? custom,
    MessageResolver? messages,
  ]) : _validators = {
          ...builtinValidators(messages),
          if (custom != null)
            for (final entry in custom.entries)
              entry.key: _withoutContext(entry.value),
        };

  /// Runs the validator for [definition]; returns null when valid, or when no
  /// validator is registered for the type (unknown types are ignored so that
  /// plugin types degrade gracefully).
  String? run(
    Object? value,
    ValidatorDefinition definition, [
    ValidationContext? context,
  ]) =>
      _validators[definition.type]
          ?.call(value, definition, context ?? const ValidationContext());

  /// Registers (or replaces) a validator implementation.
  void register(String type, ValidatorFn fn) =>
      _validators[type] = _withoutContext(fn);

  /// Registers a validator that also sees the rest of the record
  /// ([ValidationContext]) — cross-field rules of your own.
  void registerWithContext(String type, ContextValidatorFn fn) =>
      _validators[type] = fn;

  bool contains(String type) => _validators.containsKey(type);

  /// アプリが足した検証の名前だけ（組み込みは除く）。実行時に「何を登録したか」を
  /// 吐くために使う（`registrySnapshot`）。組み込みは突き合わせ側が知っているので、
  /// ここに混ぜると一覧が無駄に太り、組み込みが増えるたびに古くなる。
  List<String> get customKeys {
    final builtin = builtinValidators().keys.toSet();
    return [
      for (final key in _validators.keys)
        if (!builtin.contains(key)) key,
    ]..sort();
  }
}
